design-proposals/klip-32-sql-testing-tool.md

```diff
@@ -0,0 +1,238 @@
+```sql
+ASSERT DATA source_name [OFFSET offset] ( { column_name } [, ...] ) VALUES ( value [, ...] );
+```
+
+The `ASSERT NO DATA` statement will allow users to ensure that no more data exists in a specified
+source:
+```sql
+ASSERT NO DATA source_name;
+```
+
+The `ASSERT (TABLE | STREAM)` statement asserts that the given stream or table has the specified
+columns and physical properties.
+```sql
+ASSERT (STREAM | TABLE) source_name ( { column_name data_type [[PRIMARY] KEY] } [, ...] )
+    WITH (property_name = expression [, ...] );
+```
+
+#### Considered Alternative
+
+It is possible to use a meta directive (see below) in order to drive asserts as well. This was
+considered and rejected because it would require additional parsing and more testing-specific code
+to convert to expressions and pass into the codegen, which diverges from the motivating principle 
+of minimizing testing-only code.
+
+### Meta Directives
+
+Some functionality cannot be addressed in standard SQL syntax and will instead be supported using
+meta directives. Some examples of these directives:
+
+- The `--@expected.error` and `--@expected.message` directives will indicate that the below test 
+    expects an exception to be thrown with the specified message and type.
+- The `--@topic.denylist` directive will ensure that any topics matching the denylist will not
+    be present at the end of the test execution.
+    
+### Language Gaps
+
+There are some features that we need to implement in ksqlDB's SQL language in order to put YATT
+on par with the existing testing tools. Some of those are outlined here:
+
+- _Windowed Keys_: when we have structured key support, we can `ASSERT DATA` and just provide the
+    struct as the value. Until then, windowed keys will either be unsupported or stringified.
+- _Tombstones_: we will implement `DELETE FROM table WHERE key = value` syntax to allow tombstones
+    to be inserted into tables, both for ksqlDB and the testing tool.
+- _Null Values for Streams_: insert `null` values into streams will not be supported for the first
+    iteration of YATT. Eventually, if we want to support this, we can add a directive like
+    `--@null.value: topic key` to produce a `null` valued record into a specific topic/source.
+- _Other Unsupported Inserts_: there are other types of inserts that ksqlDB doesn't allow for at the
+    moment (e.g. `enum` support or binary formats). This is only somewhat a regression from the 
+    existing QTT tests, so we may or may not decide to support it in YATT.
+    
+### Error Reporting & UX
+
+One of the bigger concerns with the existing testing tool is that it's helpful in letting
+us know _that_ an error happened, but not so much _why_ it happened or even at times _what_
+exactly happened. To avoid this pitfall, the errors will include:
+
+- the assert statement that failed
+- the actual data, and the expected data
+- the ability to add `PRINT` statements to debug the issue further by printing topic contents
```

Comment:
The parse outputs location information when parsing the SQL. So we can also link to the line within the test file that failed, much like QTT does today.  This can save a LOT of time when investigation test failures.

Can we add this to the list please?